README.md
# Bailongma

> TDengine adatper for Prometheus, and others will happen soon.

For simple use case, just run:

```sh
bailongma
```

For all options, type `--help` for details.

Long usage here:

```sh
TDengine adapter for prometheus

USAGE:
    bailongma [OPTIONS]

FLAGS:
        --help
            Prints help information

    -V, --version
            Prints version information


OPTIONS:
    -c, --chunk-size <chunk-size>
            Sql chunk size.
            
            The larger your table column size is, the small chunk should be setted. [default: 600]

    -h, --host <host>
            TDengine host IP or hostname [default: localhost]

    -l, --level <level>
            Debug level [default: info]

    -L, --listen <listen>
            Listen to an specific ip and port [default: 0.0.0.0:10203]

    -C, --max-connections <max-connections>
            Max TDengine connections
            
            - in concurrent cases, use max as 50000 - for common use, set it as 5000 [default:
            50000]

    -M, --max-memory <max-memory>
            Max memroy, unit: GB [default: 50]

    -P, --password <password>
            TDengine password [default: taosdata]

    -p, --port <port>
            TDengine server port [default: 6030]

    -u, --user <user>
            TDengine user [default: root]

    -w, --workers <workers>
            Thread works for web request [default: 10]
```

Example:

```sh
bailongma -h tdengine -p 6030 -u root -P taospass --listen 0.0.0.0:10101
```

It will listen to port `10101`(default is `10230`).

Configure in `prometheus.yml`:

```yaml
remote_write:
  - url: "localhost:10101/adapters/prometheus/write"
remote_read:
  - url: "localhost:10101/adapters/prometheus/read"
```

The default database is `prometheus`, use query option `database` to modify this, configuration file is like:

```yaml
remote_write:
  - url: "localhost:10101/adapters/prometheus/write?database=prom1"
remote_read:
  - url: "localhost:10101/adapters/prometheus/read?database=prom1"
```

## Build and Install

```sh
cargo build
cargo install --path .
```

## blm-bench-prom

`cargo build` will also produce a benchmark tool.

```sh
blm-bench-prom http://127.0.0.1:10230/adapters/prometheus/test5 \
  --points 10 \
  --metrics 10 \
  --interval 1000 \
  --chunks 1000 \
  --samples 100
```


CHANGELOG.md
<a name="v0.2.4"></a>
## v0.2.4 (2022-01-05)


#### Bug Fixes

* **[TD-12729:**  fix tag value escape error ([aa8f5291](aa8f5291))

#### Documentation

*   relicense to MIT ([99d27145](99d27145))



<a name="v0.2.3"></a>
## v0.2.3 (2021-06-08)


#### Utils

* **blm-bench-prom:**  support table check in CI ([5c83a0d1](5c83a0d1))



<a name="v0.2.2"></a>
## v0.2.2 (2021-06-08)


#### Bug Fixes

* **remote-read:**  fix tag name not match for t_ prefix ([4dc58df3](4dc58df3))

#### Utils

*   add changelog section Utils ([4d86fedd](4d86fedd))
* **clog:**  add enhancement changelog section ([e96c5bac](e96c5bac))
* **prom-decode:**  support read/write request ([047396f1](047396f1))

#### Enhancement

* **prom:**  use different prefix for failed r/w requests ([bb3019c3](bb3019c3))



<a name="v0.2.1"></a>
## v0.2.1 (2021-06-07)


#### Bug Fixes

* **[TD-4607]:**  fix prom remote read results labels error ([c85a9cc8](c85a9cc8))

#### CI

*   add blm-bench-prom to release ([ad309aaa](ad309aaa))



<a name="v0.2.0"></a>
## v0.2.0 (2021-06-07)


#### Features

* **[TD-4607]:**  support regex match/not-match for promql ([e03be78a](e03be78a))

#### Bug Fixes

*   improve error log in case of prom data error ([4bf725f0](4bf725f0))



<a name="v0.1.1"></a>
## v0.1.1 (2021-06-04)


#### CI

*   use nightly toolchain to enable backtrace ([b6dd4e2b](b6dd4e2b))

#### Bug Fixes

*   fix sql syntax error if value contains quote ([645ed56a](645ed56a))
* **Prometheus:**  use escape_default for tag value ([a38d5041](a38d5041))



<a name="v0.1.0"></a>
## v0.1.0 (2021-06-02)


#### Bug Fixes

*   use actix arbiter instead of tokio spawn ([30a42b15](30a42b15))
*   set response status as Accepted (202) ([c1413326](c1413326))
*   max_connections not work ([7e87dfe4](7e87dfe4))
*   lowercase all tag names ([3602c492](3602c492))
*   remove rwlock ([bf3bd26f](bf3bd26f))
*   fix pool size too small ([5bfddb03](5bfddb03))
*   fix space lost in insert batch ([e131a888](e131a888))

#### Features

*   memory-efficient implements ([8d294c88](8d294c88))
*   use backtrace feature to compile in stable ([3d4d7a13](3d4d7a13))
*   support early reply to prometheus ([76570a2e](76570a2e))
*   add benchmark tool for prometheus ([cd368f1b](cd368f1b))
*   split rest feature to bailongma ([43fb09c1](43fb09c1))
*   add chunk size option to reduce error ([00e1d4d0](00e1d4d0))
*   use rest api for memory leak case ([10f8c55f](10f8c55f))
*   support config options in command line ([bed3d65b](bed3d65b))
* **CHANGELOG:**  auto generate CHANGELONG.md by clog-cli ([bbe4e75a](bbe4e75a))
* **[TD-4496]:**  add build step for bailongma ([9b132b3d](9b132b3d))





src/main.rs
use std::{
    collections::{BTreeMap, BTreeSet},
    ops::Deref,
    sync::{Arc, Mutex},
    time::Duration,
};

use actix_web::{
    middleware::Logger,
    post,
    web::{self, Bytes},
    App, HttpResponse, HttpServer, Result as WebResult,
};
use anyhow::Result;
use clap::Parser;
use dashmap::{DashMap, DashSet};

use log::*;
use prost::Message;

use libtaos::{self as taos, Taos, TaosCfgBuilder, TaosCode};

// pub mod protos;
pub mod utils;

use bailongma::*;
use taos::TaosError;
use utils::md5sum;

//use protos::WriteRequest;
fn table_name_escape(name: &str) -> String {
    let s = name
        .replace(":", "_")
        .replace(".", "_")
        .replace("-", "_")
        .replace(" ", "_")
        .to_lowercase();
    if s.len() > 190 {
        s.split_at(190).0.to_string()
    } else {
        s
    }
}
fn tag_name_escape(name: &str) -> String {
    name.replace(":", "_")
        .replace(".", "_")
        .replace("-", "_")
        .to_lowercase()
}

fn tag_value_escape(value: &str) -> String {
    value.replace("\"", "\\\"")
}

async fn handle_stable_schema<'prom>(
    _state: &AppState,
    taos: &Taos,
    database: &str,
    timeseries: &'prom TimeSeries,
) -> Result<()> {
    use itertools::Itertools;
    debug!("handle stable start");
    let (name, labels): (_, Vec<_>) = timeseries
        .labels
        .iter()
        .partition(|label| label.name == "__name__");

    // label __name__ should exist.
    assert!(name.len() == 1);

    // get metrics name
    let metrics_name = &name[0].value;
    let stable_name = table_name_escape(metrics_name);

    let mut schema = taos
        .describe(&format!("{}.{}", database, &stable_name))
        .await;
    trace!("schema: {:?}", &schema);
    //let _ = state.create_table_lock.lock().unwrap();
    if let Err(taos::Error::RawTaosError(taos::TaosError { code, ref err })) = schema {
        //let _ = state.create_table_lock.lock().unwrap();
        match code {
            taos::TaosCode::MndInvalidTableName => {
                // create super table
                let sql = format!(
                    "create stable if not exists {}.{} (ts timestamp, value double) tags (taghash binary({}), {})",
                    database,
                    stable_name,
                    34, // taghash length
                    labels
                        .iter()
                        .map(|label| { format!("t_{} binary({})", tag_name_escape(&label.name), 128) })  // TODO: default binary length is 128 
                        .join(", ")
                );
                trace!("exec sql: {}", &sql);
                taos.exec(&sql).await?;
                schema = taos
                    .describe(&format!("{}.{}", database, &stable_name))
                    .await;
            }
            taos::TaosCode::MndDbNotSelected => {
                // create database
                let sql = format!("create database if not exists {}", database);
                dbg!(taos.exec(&sql).await?);
                // create super table
                let sql = format!(
                    "create stable if not exists {}.{} (ts timestamp, value double) tags (taghash binary({}), {})",
                    database,
                    stable_name,
                    34, // taghash length
                    labels
                        .iter()
                        .map(|label| { format!("t_{} binary({})", tag_name_escape(&label.name), 128) })  // TODO: default binary length is 128 
                        .join(", ")
                );
                trace!("exec sql: {}", &sql);
                taos.exec(&sql).await?;
                schema = taos
                    .describe(&format!("{}.{}", database, &stable_name))
                    .await;
            }
            _ => {
                error!("error: TaosError {{ code: {}, err: {} }}", code, err);
                return Err(taos::Error::RawTaosError(TaosError {
                    code,
                    err: err.clone(),
                })
                .into());
            }
        }
    }

    let schema = schema.unwrap();
    use std::iter::FromIterator;
    let fields = BTreeSet::from_iter(schema.names().into_iter());

    let mut tagmap = BTreeMap::new();
    for label in &labels {
        if !fields.contains(&format!("t_{}", tag_name_escape(&label.name))) {
            let sql = format!(
                "alter stable {}.{} add tag t_{} binary({})",
                database,
                stable_name,
                tag_name_escape(&label.name),
                128
            );
            trace!("add tag {} for stable {}: {}", label.name, stable_name, sql);
            let res = taos.exec(&sql).await;
            if let Err(err) = res {
                match &err {
                    taos::Error::RawTaosError(TaosError { code, err }) => match code {
                        TaosCode::MndFieldAlreayExist => {}
                        _ => {
                            anyhow::bail!(err.clone());
                        }
                    },
                    _ => {}
                }
            }
        }
        tagmap.insert(&label.name, &label.value);
    }

    let tag_values = labels.iter().map(|label| &label.value).join("");
    let table_name = format!("{}{}", metrics_name, tag_values);
    let table_name = format!("md5_{}", md5sum(table_name.as_bytes()));
    let taghash = md5sum(tagmap.values().join("").as_bytes());

    // create sub table;
    // FIXME: It's better to keep a table exist set.
    let sql = format!(
        "create table if not exists {}.{} using {}.{} (taghash,{}) tags(\"{}\",{})",
        database,
        table_name,
        database,
        stable_name,
        tagmap
            .keys()
            .map(|v| format!("t_{}", tag_name_escape(&v)))
            .join(","),
        taghash,
        tagmap
            .values()
            .map(|value| {
                if value.len() < 127 {
                    format!("\"{}\"", tag_value_escape(value))
                } else {
                    format!("\"{}\"", tag_value_escape(&value[0..127]))
                }
            })
            .join(",")
    );
    debug!("created table {}.{}", database, table_name);
    trace!("create table with sql: {}", sql);
    if let Err(taos::Error::RawTaosError(TaosError { code, err })) = taos.exec(&sql).await {
        if err.contains("tag value too long") {
            error!("tag value too long: {}", sql);
        } else {
            anyhow::bail!(err.clone());
        }
    }
    debug!("handle stable done");
    Ok(())
}
async fn handle_table_schema(
    state: &AppState,
    taos: &Taos,
    database: &str,
    req: &WriteRequest,
) -> Result<()> {
    use futures::stream::{iter, StreamExt};
    let stream = iter(req.timeseries.iter());
    let res = stream
        .then(|ts| async move { handle_stable_schema(state, taos, database, ts).await })
        .collect::<Vec<_>>()
        .await;
    for i in res {
        let _ = i?;
    }
    // let stream = iter(req.timeseries.iter().map(|ts| {
    // handle_stable_schema(state, taos, database, ts)
    // }));
    // let vec: Vec<_> = TryStreamExt::try_collect(stream).await;
    // for ts in &req.timeseries {
    //     // handle stable
    //     handle_stable_schema(state, taos, database, ts).await?;
    // }
    debug!("handle table schema done");
    Ok(())
}
async fn write_tdengine_from_prometheus(
    state: &AppState,
    database: &str,
    req: &WriteRequest,
) -> Result<()> {
    use itertools::Itertools;
    debug!("Write tdengine from prometheus write request");
    let taos = state.pool.get()?;
    let taos = taos.deref();
    // build insert sql
    let chunks = req
        .timeseries
        .iter()
        .map(|ts| {
            let (name, labels): (_, Vec<_>) =
                ts.labels.iter().partition(|label| label.name == "__name__");
            // label __name__ should exist.
            assert!(name.len() == 1);

            // get metrics name
            let metrics_name = &name[0].value;
            let tag_values = labels.iter().map(|label| &label.value).join("");
            let table_name = format!("{}{}", metrics_name, tag_values);
            let table_name = format!("{}.md5_{}", database, md5sum(table_name.as_bytes()));
            ts.samples.iter().map(move |sample| match sample.value {
                Some(value) if value.is_nan() => {
                    format!(" {} values ({}, NULL)", table_name, sample.timestamp)
                }
                None => format!(" {} values ({}, NULL)", table_name, sample.timestamp),
                Some(value) => format!(" {} values ({}, {})", table_name, sample.timestamp, value),
            })
        })
        .flatten()
        .chunks(state.opts.chunk_size)
        .into_iter()
        .map(|mut chunk| chunk.join(""))
        .collect_vec();

    for chunk in chunks {
        let sql = format!("insert into {}", chunk);
        debug!("chunk sql length is {}", sql.len());

        if let Err(err) = taos.query(&sql).await {
            match err {
                taos::Error::RawTaosError(err) => match err.code {
                    TaosCode::MndDbNotSelected | TaosCode::MndInvalidTableName => {
                        handle_table_schema(&state, taos, database, &req).await?;
                        taos.query(&sql).await?;
                    }
                    code => {
                        warn!("insert into tdengine error: [{}]{}", code, err);
                    }
                },
                err => {
                    error!("error with query [{}]: {}", sql.len(), err);
                }
            }
        }
    }

    Ok(())
}

#[post("/adapters/prometheus/write")]
async fn prometheus(
    state: web::Data<Arc<AppState>>,
    web::Query(options): web::Query<PrometheusOptions>,
    bytes: Bytes,
) -> WebResult<HttpResponse> {
    let bytes = bytes.deref();

    info!("recieved {} bytes from prometheus", bytes.len());
    let database = options.database;
    let database = database.unwrap_or("prometheus".to_string());

    let mut decoder = snap::raw::Decoder::new();
    let decompressed = decoder
        .decompress_vec(bytes)
        .map_err(|_| actix_web::error::ErrorNotAcceptable("bad snappy stream"))?;

    let write_request = WriteRequest::decode(&mut decompressed.as_ref()).map_err(|prost_err| {
        // decompressed.len();
        let err = "bad prometheus write request: deserializing error";
        error!(
            "{}, protolens: {}, raw error: {:?}",
            err,
            decompressed.len(),
            prost_err
        );
        actix_web::error::ErrorNotAcceptable(err)
    })?;
    drop(decompressed); // drop decompressed data, it'll not be used after

    // std::fs::write(format!("prom-failed-{}.json", md5sum(bytes)), serde_json::to_string(&write_request).unwrap())?;

    // write tdengine, retry max 10 times if error.
    for _i in 0..10i32 {
        let res = write_tdengine_from_prometheus(&state, &database, &write_request).await;
        if let Err(err) = res {
            warn!(
                "write tdengine error : {}\nbacktraces: {}",
                // i,
                err,
                err.backtrace()
            );
            tokio::time::sleep(Duration::from_millis(100)).await;
        } else {
            return Ok(HttpResponse::Ok().finish());
        }
    }

    // if not success, write data and save it to persistent storage.
    error!("failed with retries, the data will be lost");
    std::fs::write(format!("prom-failed-write-{}.snappy", md5sum(bytes)), bytes)?;
    // });
    // body is loaded, now we can deserialize
    Ok(HttpResponse::InternalServerError().finish())
}

#[derive(Debug, serde::Deserialize)]
struct PrometheusOptions {
    database: Option<String>,
}
#[post("/adapters/prometheus/read")]
async fn prometheus_read_handler(
    state: web::Data<Arc<AppState>>,
    web::Query(options): web::Query<PrometheusOptions>,
    bytes: Bytes,
) -> WebResult<HttpResponse> {
    let bytes = bytes.deref();

    info!("recieved {} bytes from prometheus", bytes.len());

    let database = options.database;
    let database = database.unwrap_or("prometheus".to_string());
    // let database = database.to_string();

    let mut decoder = snap::raw::Decoder::new();
    let decompressed = decoder
        .decompress_vec(bytes)
        .map_err(|_| actix_web::error::ErrorNotAcceptable("bad snappy stream"))?;

    let read_request = ReadRequest::decode(&mut decompressed.as_ref()).map_err(|_| {
        actix_web::error::ErrorNotAcceptable("bad prometheus read request: deserializing error")
    })?;
    drop(decompressed); // drop decompressed data, it'll not be used after
    let taos = state.pool.get().expect("get connection from pool");
    let taos = taos.deref();
    for _i in 0..10i32 {
        let res = prometheus_read(taos, &database, &read_request).await;
        if let Err(err) = res {
            warn!("read tdengine error : {}", err,);
            tokio::time::sleep(Duration::from_millis(100)).await;
        } else {
            let res = res.unwrap();
            let mut buf: Vec<u8> = vec![];
            let _ = res.encode(&mut buf).map_err(|_| {
                actix_web::error::ErrorNotAcceptable("failed encode protobuf message")
            })?;
            let mut encoder = snap::raw::Encoder::new();
            let compressed = encoder.compress_vec(&buf).map_err(|_| {
                actix_web::error::ErrorNotAcceptable("failed to compress with snappy method")
            })?;
            return Ok(HttpResponse::Ok().body(compressed));
        }
    }

    // if not success, write data and save it to persistent storage.
    error!("failed with retries, the data will be lost");
    std::fs::write(format!("prom-failed-read-{}.snappy", md5sum(bytes)), bytes)?;
    // });
    // body is loaded, now we can deserialize
    Ok(HttpResponse::InternalServerError().finish())
}

/// TDengine adapter for prometheus.
#[derive(Debug, Clone, Parser)]
#[clap(setting = clap::AppSettings::ColoredHelp)]
#[clap(version, author)]
struct Opts {
    /// Debug level
    #[clap(short, long, default_value = "info")]
    level: log::LevelFilter,
    /// TDengine host IP or hostname.
    #[clap(short, long, default_value = "localhost")]
    host: String,
    /// TDengine server port
    #[clap(short, long, default_value = "6030")]
    port: u16,
    /// TDengine user
    #[clap(short, long, default_value = "root")]
    user: String,
    /// TDengine password
    #[clap(short = 'P', long, default_value = "taosdata")]
    password: String,
    /// Listen to an specific ip and port
    #[clap(short = 'L', long, default_value = "0.0.0.0:10203")]
    listen: String,
    /// Thread works for web request
    #[clap(short, long, default_value = "10")]
    workers: usize,
    /// Sql chunk size.
    ///
    /// The larger your table column size is, the small chunk should be setted.
    #[clap(short, long, default_value = "600")]
    chunk_size: usize,
    /// Max TDengine connections
    ///
    ///   - in concurrent cases, use max as 50000
    ///   - for common use, set it as 5000
    #[clap(short = 'C', long, default_value = "500")]
    max_connections: u32,
    /// Max memroy, unit: GB
    #[clap(short = 'M', long, default_value = "50")]
    max_memory: u64,

    /// Tag data type
    #[clap(short = 't', long, default_value = "binary")]
    tag_type: String,
}

#[derive(Debug, Default)]
pub struct Tables(DashSet<String>);

impl Tables {
    pub fn exist(&self, name: &str) -> bool {
        self.0.contains(name)
    }
    pub fn add_table(&mut self, name: impl Into<String>) {
        self.0.insert(name.into());
    }
}

type StableHandler = DashMap<String, (DashSet<String>, DashMap<String, Tables>)>;
#[derive(Debug, Default)]
pub struct DatabasesHandler(
    DashMap<String, DashMap<String, (DashSet<String>, DashMap<String, Tables>)>>,
);

impl DatabasesHandler {
    pub fn new() -> Self {
        Default::default()
    }

    pub fn assert_database<'a>(
        &'a self,
        database: &str,
    ) -> dashmap::mapref::one::Ref<String, StableHandler> {
        match self.0.get(database) {
            Some(st) => st,
            None => {
                self.0.insert(database.to_string(), Default::default());
                self.0.get(database).unwrap()
            }
        }
    }
}
#[derive(Debug)]
pub struct AppState {
    opts: Opts,
    pool: taos::TaosPool,
    create_table_lock: Mutex<i32>,
    tables: DatabasesHandler,
    max_memory: u64,
}

#[actix_web::main]
async fn main() -> Result<()> {
    let opts: Opts = Opts::parse();

    std::env::set_var(
        "RUST_LOG",
        format!(
            "actix_web=info,bailongma={level},main={level}",
            level = opts.level.to_string()
        ),
    );
    // fern::Dispatch::new()
    //         .level(opts.level)
    //         .chain(std::io::stdout())
    //         .chain(fern::log_file("output.log")?)
    //         .apply()?;
    env_logger::init();
    //dbg!(&opts);
    //let create_table_lock = Arc::new(Mutex::new(0));
    let taos_cfg = TaosCfgBuilder::default()
        .ip(&opts.host)
        .user(&opts.user)
        .pass(&opts.password)
        .db("log")
        .port(opts.port)
        .build()
        .expect("ToasCfg builder error");
    let taos_pool = r2d2::Pool::builder()
        .max_size(opts.max_connections)
        .test_on_check_out(false)
        .connection_timeout(Duration::from_secs(500))
        .max_lifetime(Some(Duration::from_secs(600)))
        .idle_timeout(Some(Duration::from_secs(300)))
        .build(taos_cfg)?;
    let workers = opts.workers;
    let listen = opts.listen.clone();

    let max_memory = opts.max_memory;
    let state = Arc::new(AppState {
        opts,
        pool: taos_pool.clone(),
        create_table_lock: Default::default(),
        tables: Default::default(),
        max_memory: max_memory * 1024 * 1024,
    });
    let server = HttpServer::new(move || {
        App::new()
            .data(state.clone())
            .wrap(Logger::default())
            .service(prometheus)
            .service(prometheus_read_handler)
    })
    .workers(workers)
    .bind(&listen)?
    .run();
    info!("start server, listen on {}", listen);

    server.await?;
    Ok(())
}
